app/forms/upload_form.rb

```diff
@@ -0,0 +1,51 @@
+# frozen_string_literal: true
+
+require 'csv'
+
+CSV_CONTENT_TYPE = 'text/csv'
+CSV_COUNT_MIN = 1
+CSV_COUNT_MAX = 1000
+
+class UploadForm
+  include ActiveModel::Validations
+
+  attr_reader :search_engine, :csv_file
+
+  validates :search_engine, presence: true
+  validates :search_engine, inclusion: { in: SearchResult.search_engines.keys }
+  validates :csv_file, presence: true
+
+  validate :valid_content_type, if: -> { csv_file.present? }
+  validate :valid_count, if: -> { csv_file.present? }
+
+  def initialize(search_engine:, csv_file:)
+    @search_engine = search_engine
+    @csv_file = csv_file
+  end
+
+  def valid_content_type
+    return if csv_file.content_type == CSV_CONTENT_TYPE
+
+    add_error :invalid_extension
+  end
+
+  def valid_count
+    return if CSV.read(csv_file.tempfile).count.between?(CSV_COUNT_MIN, CSV_COUNT_MAX)
+
+    add_error :invalid_count
+  end
+
+  def add_error(type)
+    errors.add(:base, I18n.t("activemodel.errors.models.search_result.attributes.csv_file.#{type}"))
+  end
```

Comment:
You don't need to get the error message with this long line. To add a custom error to an attribute, you just need:
```suggestion
  def add_error(type)
    errors.add(:csv_file, type)
  end
```
And in the `activemodel.yml`:
```yml
en:
  activemodel:
    errors:
      models:
        upload_form:
          attributes:
            csv_file:
              blank: "CSV file is required."
              invalid_extension: "CSV file is invalid, must have .csv extension."
              invalid_count: "CSV file is invalid, must have 1-1000 keywords."
```